### PR TITLE
Add RedBox crash logging to Console

### PR DESCRIPTION
**Pull request by HeyImChris**
#### Please select one of the following
- [ ] I am removing an existing difference between facebook/react-native and microsoft/react-native-macos :thumbsup:
- [ ] I am cherry-picking a change from Facebook's react-native into microsoft/react-native-macos :thumbsup:
- [ ] I am making a fix / change for the macOS implementation of react-native
- [X] I am making a change required for Microsoft usage of react-native

## Summary

We have some automation that tracks crashes and pulls the entire Console logs, however without this PR the exception logs aren't reported to the Console app as well as /usr/bin/logs.

With this in place, anyone can pull the logs and see both the JS and Native error message and stacks. Note that the most recent commits move this change out of RCTRedBox as that won't hit during production builds with RedBoxes disabled. I moved the logging to RCTExceptionsManager which should work in production

## Changelog

[Apple] [Bug] - Redbox console logging

## Test Plan

I forced both native and JS crashes in the RNTester app and found the appropriate crash logs and stacks in the macOS Console app.

### CODE REVIEW

File: React/CoreModules/RCTRedBox.mm [hunk 1]
```diff
@@ -968,6 +969,10 @@ - (void)showErrorMessage:(NSString *)message
 
     RCTErrorInfo *errorInfo = [[RCTErrorInfo alloc] initWithErrorMessage:message stack:stack];
     errorInfo = [self _customizeError:errorInfo];
+    
+    // TODO:(macOS GH#774) Explicitly log the error message to the Apple Console so production logs can be retrieved
+    os_log_error(OS_LOG_DEFAULT, "React-Native RedBox Error Message:\n%@ \nReact-Native RedBox Stack:\n%@", errorInfo.errorMessage, errorInfo.stack);
```
Comment (harrieshin) [inline React/CoreModules/RCTRedBox.mm:974]: this entire class is only for non production so its ok not to have any other build stage #ifdefs ?

**Review by harrieshin — APPROVED** [review]

**Comment by HeyImChris** [comment]
Putting this review on hold as it might not be enough for our purposes. The `os_log` seems to only get reported to the Console app when Console is running which won't suffice for our ship builds.

**Comment by HeyImChris** [comment]
> Putting this review on hold as it might not be enough for our purposes. The `os_log` seems to only get reported to the Console app when Console is running which won't suffice for our ship builds.

Got it sorted out. This reports the logs to usr/bin/logs so automation should be able to pick it up

File: React/CoreModules/RCTRedBox.mm (hunk 1 above)
Comment (HeyImChris) [reply React/CoreModules/RCTRedBox.mm]: Yea I noticed this too. I moved the error to a location that should hit production

**Comment by acoates-ms** [comment]
Shouldn't this be something that an app adds to their configuration of the instance, rather than something that should be hard coded?

Can't the app use the updateJSExceptionWithMessage delegate or something to add their own behavior here?

**Comment by HeyImChris** [comment]
> Shouldn't this be something that an app adds to their configuration of the instance, rather than something that should be hard coded?
> 
> Can't the app use the updateJSExceptionWithMessage delegate or something to add their own behavior here?

At least in Office we want these logs for every app/SDX. An app can add their own behavior here but if they all need it, why not add it in one centralized location?

**Comment by acoates-ms** [comment]
Then office should do that in a shared config within office.  That way for instance the logic could be updated to only log in automation (which would be office specific code).

**Comment by HeyImChris** [comment]
> Then office should do that in a shared config within office. That way for instance the logic could be updated to only log in automation (which would be office specific code).

We want the logs for production too (not just automation). There's nothing Office-specific about this- it's something others outside of Office can leverage too

File: React/CoreModules/RCTRedBox.mm [hunk 2]
```diff
@@ -29,7 +29,7 @@
 
 @class RCTRedBoxWindow;
 
-#if !TARGET_OS_OSX // TODO(macOS GH#774)
+#if !TARGET_OS_OSX // TODO:(macOS GH#774)
```
Comment (Saadnajmi) [inline React/CoreModules/RCTRedBox.mm:32]: Is there a new precedent for "TODO:" vs "TODO" ?

Comment (HeyImChris) [reply React/CoreModules/RCTRedBox.mm:32]: It doesn't matter much but using the colon then Xcode recognizes it

**Pull request closed without merging**